File: Lecture05-DependencyInversionPrinciple/Java/spellcheck/SpellingChecker.java

package spellcheck;

import java.util.*;
import java.net.*;
import java.io.*;


public class SpellingChecker {

	public SortedMap<String, Integer> check(URL url) throws IOException {

		// download the document content
		//
		URLFetcher fetcher = new URLFetcher();
		String content = fetcher.fetch(url);

		// extract words from the content
		//
		WordExtractor extractor = new WordExtractor();
		List<String> words = extractor.extract(content);

		// find spelling mistakes
		//
		Dictionary dictionary = new Dictionary("dict.txt");
		SortedMap<String, Integer> mistakes = new TreeMap<String, Integer>();
		
		Iterator<String> it = words.iterator();
		while (it.hasNext()) {
			String word = it.next();
			if (!dictionary.isValidWord(word)) {
				if (mistakes.containsKey(word)) {
					int oldCount = mistakes.get(word);
					mistakes.put(word, (Integer)(oldCount + 1));
				}
				else {
					mistakes.put(word, (Integer)1);
				}
			}
		}

		return mistakes;
	}

}

File: Lecture05-DependencyInversionPrinciple/Typescript/src/SpellingChecker.ts
import { URLFetcher } from './URLFetcher';
import { WordExtractor } from './WordExtractor';
import { Dictionary } from './Dictionary';


export class SpellingChecker {

	public check(url: string): Promise<Object> {
	
		return new Promise<Object>((resolve, reject) => {
					
			// download the document content
			let fetcher: URLFetcher = new URLFetcher();
			fetcher.fetch(url)
				.then((content: string) => {
				
					// extract words from the content
					let extractor: WordExtractor = new WordExtractor();
					let words: string[] = extractor.extract(content);
					
					// find spelling mistakes
					let dictionary: Dictionary = new Dictionary();
					dictionary.load("dict.txt")
						.then((loadResult: boolean) => {
						
							let mistakes: Object = {};
							
							words.forEach((word: string) => {
								if (!dictionary.isValidWord(word)) {
									if (mistakes[word] !== undefined) {
										mistakes[word] = mistakes[word] + 1;
									}
									else {
										mistakes[word] = 1;
									}
								}
							});
							
							let sortedMistakes: Object = {};
							Object.keys(mistakes).sort().forEach((key: string) => {
								sortedMistakes[key] = mistakes[key];
							});
							
							resolve(sortedMistakes);
						})
						.catch((error: string) => {
							reject(error);
						});		
				})
				.catch((error: string) => {
					reject(error);
				});			
		});
	}
	
}
